cmake/tribits/core/package_arch/TribitsProjectImpl.cmake

```diff
@@ -321,7 +321,27 @@ MACRO(TRIBITS_PROJECT_IMPL)
   ENDIF()
 
   #
-  # N) Show final timing and end
+  # N) Generate resource spec file if applicable
+  #
+
+  IF (${PROJECT_NAME}_AUTOGENERATE_TEST_RESOURCE_FILE)
+    TRIBITS_GENERATE_RESOURCE_SPEC_FILE()
+    ADVANCED_SET(CTEST_RESOURCE_SPEC_FILE "${CMAKE_BINARY_DIR}/CMakeFiles/resources.json"
+      CACHE INTERNAL
+      "Resource spec file for CTest."
+      )
+  ELSEIF(DEFINED CACHE{CTEST_RESOURCE_SPEC_FILE})
+    GET_PROPERTY(CTEST_RESOURCE_SPEC_FILE_TYPE CACHE CTEST_RESOURCE_SPEC_FILE PROPERTY TYPE)
+    IF (CTEST_RESOURCE_SPEC_FILE_TYPE STREQUAL "INTERNAL")
+      ADVANCED_SET(CTEST_RESOURCE_SPEC_FILE ""
+        CACHE INTERNAL
+        "Resource spec file for CTest."
+        )
+    ENDIF()
+  ENDIF()
```

Comment:
Normally, `CTEST_RESOURCE_SPEC_FILE` is intended to be set by the user with a `-D` argument to `cmake`. However, since we are generating the file ourselves, it doesn't make sense to also provide the option to the user to change it, so I made it an `INTERNAL` option to tell CMake about it while hiding it from the user in `ccmake` and `cmake-gui`. If the project then gets reconfigured with `${PROJECT_NAME}_AUTOGENERATE_TEST_RESOURCE_FILE=OFF`, we check if the cache entry exists and is `INTERNAL`, meaning it was previously generated by `${PROJECT_NAME}_AUTOGENERATE_TEST_RESOURCE_FILE=ON`. If so, that means it was not manually specified by the user, so we remove it.

Reply:
Instead of this, could we not just give a default value to `CTEST_RESOURCE_SPEC_FILE` if `${PROJECT_NAME}_AUTOGENERATE_TEST_RESOURCE_FILE=ON` but not hide the cache variable `CTEST_RESOURCE_SPEC_FILE`?   If the full path given in `CTEST_RESOURCE_SPEC_FILE` is not the default value then `${PROJECT_NAME}_AUTOGENERATE_TEST_RESOURCE_FILE=ON` would be ignored with a "NOTE" printed to the screen.

You would have the following just below the definition of `${PROJECT_NAME}_AUTOGENERATE_TEST_RESOURCE_FILE` in the macro `TRIBITS_DEFINE_GLOBAL_OPTIONS_AND_DEFINE_EXTRA_REPOS()`:

```
SET(CTEST_RESOURCE_SPEC_FILE_DOC_EXTRA "")
IF (${PROJECT_NAME}_AUTOGENERATE_TEST_RESOURCE_FILE)
  SET(CTEST_RESOURCE_SPEC_FILE_DEFAULT  ${CMAKE_BINARY_DIR}/ctest_resources.json)
  IF ("${CTEST_RESOURCE_SPEC_FILE}" STREQUAL "")
    SET(CTEST_RESOURCE_SPEC_FILE_DOC_EXTRA
       "  This file is autogenerated by default since ${PROJECT_NAME}_AUTOGENERATE_TEST_RESOURCE_FILE=${${PROJECT_NAME}_AUTOGENERATE_TEST_RESOURCE_FILE}!" )
  ENDIF()
ELSE()
  SET(CTEST_RESOURCE_SPEC_FILE_DEFAULT "")
ENDIF()

ADVANCED_SET(CTEST_RESOURCE_SPEC_FILE 
  "${CTEST_RESOURCE_SPEC_FILE_DEFAULT}"
  CACHE FILEPATH
  "Resource spec file for CTest.${CTEST_RESOURCE_SPEC_FILE_DOC_EXTRA}"
  )
```

Then when you auto-generate that file in `TribitsProjectImpl.cmake` you would have:

```
IF (${PROJECT_NAME}_AUTOGENERATE_TEST_RESOURCE_FILE)
  IF ("${CTEST_RESOURCE_SPEC_FILE}" STREQUAL "${CTEST_RESOURCE_SPEC_FILE_DEFAULT}")
    TRIBITS_GENERATE_RESOURCE_SPEC_FILE()
  ELSE()
    MESSAGE("NOTE: The test resource file CTEST_RESOURCE_SPEC_FILE='${CTEST_RESOURCE_SPEC_FILE}'"
      " will not be auto-generated even through"
      " ${PROJECT_NAME}_AUTOGENERATE_TEST_RESOURCE_FILE=${${PROJECT_NAME}_AUTOGENERATE_TEST_RESOURCE_FILE}"
      " because its location does not match the default"
      " location '${CTEST_RESOURCE_SPEC_FILE_DEFAULT}'."
      "  If you want to auto-generate this file, please clear CTEST_RESOURCE_SPEC_FILE and
      " reconfigure or create that file on your own and clear"
      " ${PROJECT_NAME}_AUTOGENERATE_TEST_RESOURCE_FILE."
      )
  ENDIF()
ELSE()
  # Don't autogenerate the resource file
ENDIF()
```

That way, we don't have to hide the variable `CTEST_RESOURCE_SPEC_FILE` and we allow the documentation for this variable to describe what is going on and what the relationship is between the variables `CTEST_RESOURCE_SPEC_FILE` and `${PROJECT_NAME}_AUTOGENERATE_TEST_RESOURCE_FILE`.

How does hat sound?

NOTE: There is complex enough logic here that we should figure our how to create some automated tests for this in TirBITS to sustain this.  This may be hard to do because of the need to set `TPL_ENABLE_CUDA=ON`.  I can't set that in TriBITS automated testing but I may be able to find to mock this and test the different use cases for the generation and and usage of this file.  I will worry about how to add automated testing for this in TriBITS to make sure we can sustain this long term.  But I can do that after we get some initial testing on CUDA/GPU systems to make sure we are on the right track with all of this.

Reply:
I think that sounds like a good idea. As far as testing it, we can use a mock variable and then put the logic inside `IF (TPL_ENABLE_CUDA OR TPL_ENABLE_CUDA_MOCK)`. Does that sound good?

Reply:
> I think that sounds like a good idea. As far as testing it, we can use a mock variable and then put the logic inside IF (TPL_ENABLE_CUDA OR TPL_ENABLE_CUDA_MOCK). Does that sound good?

Right, something like that would work.  Then we can set up different scenarios and make sure that the right resource file gets generated and the right test properties are getting set.  We can work on that once I can do some testing on 'waterman'.